solarkraft/src/state/value.ts

```diff
@@ -11,133 +53,125 @@ interface Value { }
-    val: bigint
-    type = "i64"
+export type IntValue_i64 = {
+    type: "i64"
+}
 
-    isValid(): boolean {
-        return (-(2n ** 63n) <= this.val) && (this.val < (2n ** 63n))
-    }
+export function i64(v: bigint): Value {
+    return mkInt("i64", v)
 }
 
 // u128
-export class IntValue_u128 extends IntValue {
-    val: bigint
-    type = "u128"
+export type IntValue_u128 = {
+    type: "u128"
+}
 
-    isValid(): boolean {
-        return (0n <= this.val) && (this.val <= 2n ** 128n)
-    }
+export function u128(v: bigint): Value {
+    return mkInt("u128", v)
 }
 
 // i128
-export class IntValue_i128 extends IntValue {
-    val: bigint
-    type = "i128"
+export type IntValue_i128 = {
+    type: "i128"
+}
 
-    isValid(): boolean {
-        return (-(2n ** 127n) <= this.val) && (this.val < (2n ** 127n))
-    }
+export function i128(v: bigint): Value {
+    return mkInt("i128", v)
 }
 
 // true or false
-export class BoolValue implements Value {
+export type BoolValue = {
     val: boolean
+    type: "bool"
+}
 
-    constructor(v: boolean) {
-        this.val = v
-    }
+export function bool(v: boolean): Value {
+    return { type: "bool", val: v }
 }
 
 // Symbols are small efficient strings up to 32 characters in length and limited to a-z A-Z 0-9 _ that are encoded into 64-bit integers.
 // We store the string representation, and optionally the number.
 // TODO: determine _how_ the strings are encoded as numbers
-export class SymValue implements Value {
+export type SymbValue = {
     val: string
+    type: "symb"
     num?: number
+}
 
-    private regex: RegExp = /^[a-zA-Z0-9_]{0,32}$/
-
-    constructor(v: string) {
-        this.val = v
-        if (!this.regex.test(v)) {
-            throw new TypeError(`Symbols must be up to 32 alphanumeric characters or underscores, found: ${v}.`)
-        }
+export function symb(s: string): Value {
+    const obj = { type: "symb", val: s } as Value
+    if (!isValid(obj)) {
+        throw new TypeError(`Symbols must be up to 32 alphanumeric characters or underscores, found: ${s}.`)
     }
+    return obj
 }
 
 // Addresses are always length-56
-export class AddrValue implements Value {
+export type AddrValue = {
     val: string
+    type: "addr"
+}
 
-    private regex: RegExp = /^[A-Z0-9]{56}$/
-
-    constructor(v: string) {
-        this.val = v
-        if (!this.regex.test(v)) {
-            throw new TypeError(`Address must be exactly 56 uppercase alphanumeric characters, found: ${v}.`)
-        }
+export function addr(s: string): Value {
+    const obj = { type: "addr", val: s } as Value
+    if (!isValid(obj)) {
+        throw new TypeError(`Symbols must be up to 32 alphanumeric characters or underscores, found: ${s}.`)
     }
+    return obj
 }
 
 // Byte arrays (Bytes, BytesN)
 // The `len` field is present iff the length is fixed (i.e. for BytesN)
 // TODO: tests
-export class ArrValue implements Value {
+export type ArrValue = {
     val: IntValue_u32[]
```

Comment:
if we already know that we have an unsigned `u32` here, why don't we just use `number`? Most likely, you would have to use the array length in loops, which would make it much easier if we are using `number`.